### PR TITLE
Yahoo crypto Swift 4

### PR DESCRIPTION
**Pull request by jsom**

### CODE REVIEW

**Comment by oliver-was-here** [comment]
@jsom looks like this needs a rebase? Not entirely sure where I should start reviewing from -- should this be piggy-backing off of 05b1504 ?

File: TradeItIosTicketSDK2/CellReuseId.swift [hunk 1]
```diff
@@ -0,0 +1,9 @@
+enum CellReuseId: String {
```
Comment (oliver-was-here) [inline TradeItIosTicketSDK2/CellReuseId.swift]: Nice. We could probably stand to do this for all of our IDs.

File: TradeItIosTicketSDK2/TradeItLinkedBroker.swift [hunk 2]
```diff
@@ -148,23 +150,23 @@ import PromiseKit
...
                     onFailure: { errorResult in
-                        fulfill(())
+                        seal.fulfill(())
                     }
                 )
             }
         }
 
-        _ = when(resolved: promises).always {
+        _ = when(resolved: promises).done { _ in
```
Comment (oliver-was-here) [inline TradeItIosTicketSDK2/TradeItLinkedBroker.swift]: `_ =` is assigning the return value to a black-hole - yea? What's the purpose of this?

File: TradeItIosTicketSDK2/TradeItLinkedBroker.swift [hunk 3]
```diff
@@ -148,23 +150,23 @@ import PromiseKit
         cacheResult: Bool = true,
         onFinished: @escaping () -> Void
     ) {
-        let promises = accounts.filter { account in
+        let promises: [Promise<Void>] = accounts.filter { account in
             return force || (account.balance == nil && account.fxBalance == nil)
         }.map { account in
-            return Promise<Void> { fulfill, reject in
+            return Promise<Void> { (seal: Resolver<Void>) -> Void in
```
Comment (oliver-was-here) [inline TradeItIosTicketSDK2/TradeItLinkedBroker.swift]: All that's happening is PK6 creates a DTO for `fulfill` and `resolve` yea?

File: TradeItIosTicketSDK2/TradeItNumericToggleInputCell.swift [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+class TradeItNumericToggleInputCell: UITableViewCell {
+    @IBOutlet weak var textField: TradeItNumberField!
+    @IBOutlet weak var quantityTypeButton: UIButton!
+
+    var onValueUpdated: ((_ newValue: NSDecimalNumber?) -> Void)?
+    var onQuantityTypeToggled: (() -> Void)?
+
+    override func awakeFromNib() {
+        self.textField.padding = UIEdgeInsets(top: 0, left: 5, bottom: 0, right: 5)
```
Comment (oliver-was-here) [inline TradeItIosTicketSDK2/TradeItNumericToggleInputCell.swift]: Can this be done in the interface builder? Seems more consistent with our other screens/nibs if so.

File: TradeItIosTicketSDK2/TradeItBundleProvider.swift [hunk 5]
```diff
@@ -5,4 +5,56 @@
...
+
+    static func registerBrandedAccountNibCells(forTableView tableView: UITableView) {
+        let bundle = provide()
+        tableView.register(
+            UINib(nibName: "TradeItBrandedAccountNameCell", bundle: bundle),
+            forCellReuseIdentifier: "BRANDED_ACCOUNT_NAME_CELL_ID"
+        )
+    }
+
+    static func registerYahooNibCells(forTableView tableView: UITableView) {
```
Comment (oliver-was-here) [inline TradeItIosTicketSDK2/TradeItBundleProvider.swift]: This is specific to the trading screen yea? `registerYahooTradingTicketNibCells`?

File: TradeItIosTicketSDK2/TradeService.swift [hunk 6]
```diff
@@ -1,26 +1,47 @@
...
+        onSuccess: @escaping (Self.PreviewTradeResult) -> Void,
+        onFailure: @escaping (TradeItErrorResult) -> Void
+    )
+
+    func placeTrade(
+        _ data: TradeItPlaceTradeRequest,
+        withCompletionBlock completionBlock: @escaping (TradeItResult) -> Void
+    )
+
+    func answerSecurityQuestionPlaceOrder(
```
Comment (oliver-was-here) [inline TradeItIosTicketSDK2/TradeService.swift]: `...PlaceOrder` vs `...placeTrade` is there a reason for this difference in vernacular?

File: TradeItIosTicketSDK2/TradeItYahooTradePreviewViewController.swift [hunk 7]
```diff
@@ -168,7 +161,7 @@ class TradeItYahooTradePreviewViewController: TradeItYahooViewController, UITabl
     }
 
     @IBAction func actionButtonTapped(_ sender: UIButton) {
-        if self.placeOrderResult != nil {
+        if self.dataSource?.isOrderPlaced == true {
```
Comment (oliver-was-here) [inline TradeItIosTicketSDK2/TradeItYahooTradePreviewViewController.swift]: this seems a little silly. Is there a reason why we can't just `...isOrderPlaced {`?

File: TradeItIosEmsApi/TradeItInstrumentActionCapability.h [hunk 8]
```diff
@@ -0,0 +1,11 @@
+#import "TradeItInstrumentCapability.h"
+
+@protocol TradeItInstrumentActionCapability
+
+@end
+
+@interface TradeItInstrumentActionCapability : TradeItInstrumentCapability
+
+@property (nonatomic, copy, nonnull) NSArray<NSString *> *supportedOrderQuantityTypes;
```
[the remark this answers is not in the document]
Comment (jsom) [reply TradeItIosEmsApi/TradeItInstrumentActionCapability.h]: Talked in person

File: TradeItIosTicketSDK2/TradeItYahooTradePreviewViewController.swift (hunk 7 above)
Comment (jsom) [reply TradeItIosTicketSDK2/TradeItYahooTradePreviewViewController.swift]: `Value of optional type 'Bool?' not unwrapped; did you mean to use '!' or '?'?`

File: TradeItIosTicketSDK2/TradeService.swift (hunk 6 above)
Comment (jsom) [reply TradeItIosTicketSDK2/TradeService.swift]: It diverged in the past. Can update but would rather do that as a separate chore as this PR is too big.

File: TradeItIosTicketSDK2/TradeItNumericToggleInputCell.swift (hunk 4 above)
Comment (jsom) [reply TradeItIosTicketSDK2/TradeItNumericToggleInputCell.swift]: I just removed this. Seems like it was not needed.

**Pull request merged**